tests/acceptance/30_custom_promise_types/24_expireafter.cf

```diff
@@ -0,0 +1,90 @@
+##############################################################################
+#
+#  Test that expireafter works with custom promise types.
+#
+##############################################################################
+
+body common control
+{
+  inputs => { "../default.cf.sub" };
+  bundlesequence  => { default("$(this.promise_filename)") };
+  version => "1.0";
+}
+
+bundle common version_check
+{
+  classes:
+      "compatible_python_version"
+        expression => returnszero(
+          "/usr/bin/python3 -c 'import sys; assert sys.version_info >= (3,6)'",
+          "useshell"
+        );
+
+@if minimum_version(3.22)
+      "supports_expireafter"
+        expression => "any";
+@endif
+}
+
+##############################################################################
+
+bundle agent init
+{
+  files:
+      "$(this.promise_dirname)/cfengine.py"
+        copy_from => local_cp(
+          "$(this.promise_dirname)/../../../modules/promises/cfengine.py"
+        );
+      "$(G.testfile)"
+        delete => tidy;
+}
+
+##############################################################################
+
+body action background {
+  background => "true";
+}
+
+bundle agent test
+{
+  meta:
+      "description" -> { "CFE-3435" }
+        string => "Test that expireafter works with custom promise types";
+
+      "test_skip_unsupported"
+        string => "!supports_expireafter|!compatible_python_version";
+
+  commands:
+      "$(sys.cf_agent) --inform --file $(this.promise_filename).sub"
+        action => background,
+        comment => "I will get killed by the second agent",
+        handle => "unique"; # Handle makes the promise lock unique.
+      "/bin/sleep 90"
+        comment => "I will wait for the first agent to expire";
```

Comment:
Doesn't this block the next cf-agent execution (on line 64)for 90 seconds. What do you mean that you will wait for the first agent to expire?

Isn't it that the first agent gets triggered, backgrounded and in that subagent execution there is a promise to write `Hello CFEngine!` after 120 seconds. In parallel this promise starts sleeping for 90 seconds but it's not backgrounded so it blocks the next promise which runs the same policy as above. When it runs it sees the pre-existing promise, expires it, and then itself proceeds to wait 120 seconds before writing the file  at which point this parent agent proceeds to the check bundle to look and see if the expected content is present.

Given that both executions of the subagent policy are identical, how do you know that the single Hello CFEngine line is from the second agent execution and not the first and that it actually expired the first?